### PR TITLE
Issue 182: number of active maintainers

### LINKED ISSUE #182: List active maintainers
### What can be improved?

ecosyste.ms [packages API](https://packages.ecosyste.ms/docs/index.html) has data on the "active" maintainers of a repo. We could use this to define a new "active maintainers" column to the tracker.

### Version

2025-11-06

### PR DESCRIPTION
**Pull request by finozzifa**
Fixes #182 

## Summary of changes in this pull request

- In `inventory/get_stats.py`, I have introduced the function `_get_number_of_maintainers` that computes the number of the active maintainers of a repository, starting from the field "issue_metadata" > "active_maintainers" of the Package API response.
- In `inventory/get_stats.py`, I have introduced the new field `active_maintainers_count` in the `ENTRIES_TO_KEEP` list.
- In `inventory/get_stats.py`, I have included in the function `get_ecosystems_entry_data`, the call to `_get_number_of_maintainers`
- In the app code, I have updated `COLUMN_NAME_MAPPING`, `COLUMN_DTYPES`, `NUMBER_FORMAT` and `COLUMN_HELP`

I would be happy to also add the necessary unit tests for `_get_number_of_maintainers` if the time allows.  

## Further changes
@brynpickering : four unit tests were failing. I have spotted the cause in `website/pages/1_👤_Deep_Dive_-_User_Interaction_Analysis.py` and proposed a change. Let me know whether you agree with it.

## Contributor checklist

- [x] Licensing information to new or modified files has been added (SPDX header, REUSE.toml, etc.).
- [x] By contributing I agree to make my contributions available under the repository's MIT license and the data generated under the repository's CC-BY-4.0 license.
- [x] I have added myself to the list of contributors in `AUTHORS.md` (if applicable).

## Reviewer checklist

- [x] Test(s) added to cover contribution
- [x] Changelog updated

### CODE REVIEW

File: inventory/get_stats.py [hunk 1]
```diff
@@ -176,6 +179,38 @@ def _get_nested_dict_entry(
...
+        LOGGER.warning(f"Error fetching ecosyste.ms package data for {url}: {e}")
+        return 0
+
+    if not package_data or package_data == "not-found":
+        LOGGER.warning(f"Could not find ecosyste.ms package entry for {url}")
+        return 0
+
+    try:
+        active_maintainers = package_data[0]["issue_metadata"]["active_maintainers"]
+    except (IndexError, TypeError) as e:
```
Comment (brynpickering) [inline inventory/get_stats.py]: This should also cover `KeyError` (when there is "issue_metadata" but not including "active_maintainers")

File: inventory/get_stats.py [hunk 2]
```diff
@@ -176,6 +179,38 @@ def _get_nested_dict_entry(
...
 
 
+def _get_number_of_maintainers(url: str) -> int:
+    """Get the number of active maintainers for a repository from ecosyste.ms package API.
+
+    Args:
+        url (str): Repository URL.
+
+    Returns:
+        int: Number of active maintainers. Returns 0 if no entry is found or data is unavailable.
```
Comment (brynpickering) [inline inventory/get_stats.py]: I would prefer a value that is clearly not valid when the data cannot be found, e.g. None or -1. Then we can differentiate between being known as "not actively maintained" and no data being available

**Review by brynpickering — CHANGES_REQUESTED** [review]
Looks good. Very straightforward. We're not strict on testing since most of the edge cases we need to test against are to do with how the scripts respond to what the `requests` queries return and we never know what all the possibilities are there.

**Comment by brynpickering** [comment]
@finozzifa digging into the various ecosyste-ms APIs, it seems what we want for more consistent results is to use https://issues.ecosyste.ms/docs/index.html / https://issues.ecosyste.ms/api/v1/hosts/GitHub/repositories/

Since the `packages` URL only has data when the project has a package indexed somewhere (which many of them don't).

File: website/⚡️_Tool_Repository_Metrics.py [hunk 3]
```diff
@@ -36,6 +36,7 @@
     "category": "Category",
     "language": "Language",
     "license_category": "License Type",
+    "active_maintainers_count": "Maintainers",
```
Comment (brynpickering) [inline website/⚡️_Tool_Repository_Metrics.py]: Move this to just after "contributors" so that it moves to the left in the resulting table

File: inventory/get_stats.py (hunk 1 above)
Comment (finozzifa) [reply inventory/get_stats.py]: Done!

File: inventory/get_stats.py (hunk 2 above)
Comment (finozzifa) [reply inventory/get_stats.py]: Done!

**Comment by brynpickering** [comment]
@finozzifa when inspecting the generated dashboard, I see a few issues  that need rectifying:

1. tools with `-1` active maintainers should be displayed as "NaN" (i.e. no data available) to align with all other cases where we have no data for a metric. 
2. there should be a check for tools with zero _active_ maintainers that they have at least one maintainer. If they have zero maintainers of any kind then the number of active maintainers should be set to `-1` since it suggests a problem with the ecosyste.ms API.

The first issue seems to be affected by #195 since repos.ecosyste.ms now handles the `NatLabRockies` org name, but issues.ecosyste.ms does not (it still expected the `NREL` org name).

The second issue links to #218 since I see it being the case for Gitlab repos. It seems the issues.ecosyste.ms API doesn't correctly parse the GitLab contributors API so it returns empty lists for both `maintainers` and `active_maintainers`.

**Comment by finozzifa** [comment]
> @finozzifa when inspecting the generated dashboard, I see a few issues that need rectifying:
> 
>     1. tools with `-1` active maintainers should be displayed as "NaN" (i.e. no data available) to align with all other cases where we have no data for a metric.
> 
>     2. there should be a check for tools with zero _active_ maintainers that they have at least one maintainer. If they have zero maintainers of any kind then the number of active maintainers should be set to `-1` since it suggests a problem with the ecosyste.ms API.
> 
> 
> The first issue seems to be affected by #195 since repos.ecosyste.ms now handles the `NatLabRockies` org name, but issues.ecosyste.ms does not (it still expected the `NREL` org name).
> 
> The second issue links to #218 since I see it being the case for Gitlab repos. It seems the issues.ecosyste.ms API doesn't correctly parse the GitLab contributors API so it returns empty lists for both `maintainers` and `active_maintainers`.

Hi @brynpickering thanks for your comment.

`2. there should be a check for tools with zero _active_ maintainers that they have at least one maintainer. If they have zero maintainers of any kind then the number of active maintainers should be set to `-1` since it suggests a problem with the ecosyste.ms API.` I am not sure of what other check should be implemented.

For example for this [repository](https://github.com/epri-dev/opender#OpenDER), both the repos and issues ecosystem APIs return zero maintainers. Should we in this case just show `None` as well?

Thanks!

File: inventory/get_stats.py [hunk 4]
```diff
@@ -176,6 +179,60 @@ def _get_nested_dict_entry(
...
+        maintainers = issues_data["maintainers"]
+    except (KeyError, TypeError) as e:
+        LOGGER.debug(f"Could not find maintainers field for {url}: {e}")
+        return -1
+
+    if maintainers:
+        LOGGER.warning(
+            f"No active maintainers for {url}, falling back to {len(maintainers)} total maintainers"
+        )
+        return len(maintainers)
```
Comment (brynpickering) [inline inventory/get_stats.py]: I wouldn't return the number of maintainers if active maintainers is empty. This will inflate the number for repositories with no active maintainers. 

That is, if they have maintainers but none are active then the number of active maintainers should be **zero**. If they have no active maintainers _and_ no maintainers then the value should be -1.

File: website/⚡️_Tool_Repository_Metrics.py [hunk 5]
```diff
@@ -59,13 +61,15 @@
...
     "1 Month Downloads": "localized",
+    "Active Maintainers": "localized",
 }
 
 COLUMN_HELP: dict[str, str] = {
     "Created": "First ever repository commit",
     "Updated": "Most recent repository commit",
     "Stars": "Repository bookmarks",
     "Contributors": "All-time source code contributors",
+    "Active Maintainers": "Active maintainers of the source code",
```
Comment (brynpickering) [inline website/⚡️_Tool_Repository_Metrics.py]: Looking at the issues [API source code](https://github.com/ecosyste-ms/issues/blob/7e5470935a32e7a3271a4352826e76dc27b69214/app/controllers/authors_controller.rb#L38) it seems that they set maintainers as "active" if they have contributed something in the past year. So, we could update the help text to clarify this.

**Review by brynpickering — CHANGES_REQUESTED** [review]
Almost there! Just some minor changes requested

File: website/⚡️_Tool_Repository_Metrics.py (hunk 3 above)
Comment (finozzifa) [reply website/⚡️_Tool_Repository_Metrics.py]: Done!

File: website/⚡️_Tool_Repository_Metrics.py (hunk 5 above)
Comment (finozzifa) [reply website/⚡️_Tool_Repository_Metrics.py]: Thanks it makes sense. I have modified the text to 

`Maintainers currently engaged with the project, defined as those with recent contributions (past year).`

File: inventory/get_stats.py (hunk 4 above)
Comment (finozzifa) [reply inventory/get_stats.py]: Solved!

**Comment by finozzifa** [comment]
@brynpickering I solved all your comments. Thanks!

**Review by brynpickering — APPROVED** [review]
